lib/facter/util/resolvers/ffi/hostname.rb

```diff
@@ -19,12 +19,12 @@ class AddrInfo < ::FFI::Struct
 
         module Hostname
           HOST_NAME_MAX = 64
-          EAI_NONAME = 8
+          EAI_NONAME = -2
```

Comment:
I see `/usr/include/netdb.h` defines this as:

```
# define EAI_NONAME       -2    /* NAME or SERVICE is unknown.  */
```

But when I call the method with a bogus hostname I get `8`:

```
irb(main):052:0> Facter::Util::Resolvers::Ffi::Hostname.getaddrinfo('nada', '', hints.to_ptr, ret)
=> 8
```

We defined the `getaddrinfo` function as returning a signed `:int`, so there isn't a signed/unsigned mismatch... Also if I attach `gai_strerror`, the message looks correct for `8`, but not any negative number:

```
irb(main):053:0>  Facter::Util::Resolvers::Ffi::Hostname.gai_strerror(8)
=> "nodename nor servname provided, or not known"
irb(main):054:0>  Facter::Util::Resolvers::Ffi::Hostname.gai_strerror(-2)
=> "Unknown error"
```

So I'm inclined to leave it as `EAI_NONAME = 8` and maybe a note about the discrepancy?

Reply:
Very strange... I also see -2 when using jruby in puppetserver packages. Given this monkey patch to facter 4:

```ruby
require 'facter'
require 'facter/framework/core/file_loader'
require 'ffi'

module Facter
  module Util
    module Resolvers
      module Ffi
        class AddrInfo < ::FFI::Struct
          layout  :ai_flags, :int,
                  :ai_family, :int,
                  :ai_socketype, :int,
                  :ai_protocol, :int,
                  :ai_addrlen, :uint,
                  :ai_addr, :pointer,
                  :ai_canonname, :pointer,
                  :ai_next, :pointer
        end

        module Hostname
          extend ::FFI::Library
          ffi_lib ::FFI::Library::LIBC

          attach_function :getaddrinfo, %i[string string pointer pointer], :int
          attach_function :gai_strerror, %i[int], :string
        end
      end
    end
  end
end

ret = FFI::MemoryPointer.new(Facter::Util::Resolvers::Ffi::AddrInfo)

hints = Facter::Util::Resolvers::Ffi::AddrInfo.new
hints[:ai_family] = Socket::AF_UNSPEC
hints[:ai_socketype] = Socket::SOCK_STREAM
hints[:ai_flags] = Socket::AI_CANONNAME

res = Facter::Util::Resolvers::Ffi::Hostname.getaddrinfo('nada', '', hints.to_ptr, ret)
puts res
puts Facter::Util::Resolvers::Ffi::Hostname.gai_strerror(res)
```

On Redhat 8, I get:

```
[root@seismic-subtype ~]# java -version
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (build 1.8.0_282-b08)
OpenJDK 64-Bit Server VM (build 25.282-b08, mixed mode)
[root@seismic-subtype ~]# /opt/puppetlabs/server/bin/puppetserver ruby --version
jruby 9.2.14.0 (2.5.7) 2020-12-08 ebe64bafb9 OpenJDK 64-Bit Server VM 25.282-b08 on 1.8.0_282-b08 +jit [linux-x86_64]
[root@seismic-subtype ~]# /opt/puppetlabs/server/bin/puppetserver ruby getaddrinfo.rb
-2
Name or service not known
```

But using jruby 9.2.13.0 on macOS (I can't get 9.2.14.0 to install):

```
$ java -version
java version "1.8.0_162"
Java(TM) SE Runtime Environment (build 1.8.0_162-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.162-b12, mixed mode)
$ bx ruby --version
jruby 9.2.13.0 (2.5.7) 2020-08-03 9a89c94bcc Java HotSpot(TM) 64-Bit Server VM 25.162-b12 on 1.8.0_162-b12 +jit $ bx ruby getaddrinfo.rb
8
nodename nor servname provided, or not known
```

Reply:
I assume this value is `networking resolver` dependent, -2 on Linux using glibc/resolv and 8 on *BSD (including macOS) using ISC libbind

Reply:
I've had dropped the check for EAI_NONAME because CFacter only uses it to log a different message: https://github.com/puppetlabs/facter/blob/ce1f2bb4a91a1ac4ae5852091c96ae6ee3712e23/lib/src/facts/posix/networking_resolver.cc#L96